# Coursera Capstone Project

## Analyzing Neighborhood Data Toronto

##### November 2020

---

### Part 1  - Data Frame

#### Import Libraries

In [1]:
import pandas as pd
import numpy as np

#### Load and Explore the Data

In [2]:
url='https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'

In [3]:
url

'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'

In [4]:
df_list = pd.read_html(url)

In [5]:
df_list

[    Postal Code           Borough  \
 0           M1A      Not assigned   
 1           M2A      Not assigned   
 2           M3A        North York   
 3           M4A        North York   
 4           M5A  Downtown Toronto   
 ..          ...               ...   
 175         M5Z      Not assigned   
 176         M6Z      Not assigned   
 177         M7Z      Not assigned   
 178         M8Z         Etobicoke   
 179         M9Z      Not assigned   
 
                                          Neighbourhood  
 0                                         Not assigned  
 1                                         Not assigned  
 2                                            Parkwoods  
 3                                     Victoria Village  
 4                            Regent Park, Harbourfront  
 ..                                                 ...  
 175                                       Not assigned  
 176                                       Not assigned  
 177                

In [6]:
len(df_list)

3

In [7]:
df_list[0]

Postal Code           Borough  \
0           M1A      Not assigned   
1           M2A      Not assigned   
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
..          ...               ...   
175         M5Z      Not assigned   
176         M6Z      Not assigned   
177         M7Z      Not assigned   
178         M8Z         Etobicoke   
179         M9Z      Not assigned   

                                         Neighbourhood  
0                                         Not assigned  
1                                         Not assigned  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
..                                                 ...  
175                                       Not assigned  
176                                       Not assigned  
177                                       Not assigned  
178  Mimico NW, The Queensway West, South of Bloor,...  
179                                       Not assigned  

[180 rows x 3 columns]

There are 3 Dataframes in df_list. The first one contains the Data we are looking for!  

In [8]:
df = df_list[0]

In [9]:
df.head(3)

Postal Code       Borough Neighbourhood
0         M1A  Not assigned  Not assigned
1         M2A  Not assigned  Not assigned
2         M3A    North York     Parkwoods

In [10]:
df.shape

(180, 3)

###### Drop Borough 'Not Assigned'

These are the unique Borough in Toronto (including 'Not assigned')

In [11]:
df['Borough'].unique()

array(['Not assigned', 'North York', 'Downtown Toronto', 'Etobicoke',
       'Scarborough', 'East York', 'York', 'East Toronto', 'West Toronto',
       'Central Toronto', 'Mississauga'], dtype=object)

In [12]:
df['Borough'].value_counts()

Not assigned        77
North York          24
Downtown Toronto    19
Scarborough         17
Etobicoke           12
Central Toronto      9
West Toronto         6
York                 5
East Toronto         5
East York            5
Mississauga          1
Name: Borough, dtype: int64

In [13]:
df2 = df[df['Borough'] != 'Not assigned']

In [14]:
df2

Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
..          ...               ...   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                         Neighbourhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
..                                                 ...  
160      The Kingsway, Montgomery Road, Old Mill North  
165                               Church and Wellesley  
168  Business reply mail Processing Centre, South C...  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...  
178  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

In [15]:
df2.head(3)

Postal Code           Borough              Neighbourhood
2         M3A        North York                  Parkwoods
3         M4A        North York           Victoria Village
4         M5A  Downtown Toronto  Regent Park, Harbourfront

###### Reset Index

In [16]:
df2.reset_index(inplace=True)

In [17]:
df2.head(3)

index Postal Code           Borough              Neighbourhood
0      2         M3A        North York                  Parkwoods
1      3         M4A        North York           Victoria Village
2      4         M5A  Downtown Toronto  Regent Park, Harbourfront

In [18]:
df2.drop('index', axis=1, inplace=True)

/opt/conda/envs/Python-3.7-main/lib/python3.7/site-packages/pandas/core/frame.py:3997: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  errors=errors,


In [19]:
df2.head(3)

Postal Code           Borough              Neighbourhood
0         M3A        North York                  Parkwoods
1         M4A        North York           Victoria Village
2         M5A  Downtown Toronto  Regent Park, Harbourfront

###### Neighbourhood 'Not Assigned'

Below are the unique "neighbourhoods" of Toronto

In [20]:
df2['Neighbourhood'].unique()

array(['Parkwoods', 'Victoria Village', 'Regent Park, Harbourfront',
       'Lawrence Manor, Lawrence Heights',
       "Queen's Park, Ontario Provincial Government",
       'Islington Avenue, Humber Valley Village', 'Malvern, Rouge',
       'Don Mills', 'Parkview Hill, Woodbine Gardens',
       'Garden District, Ryerson', 'Glencairn',
       'West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale',
       'Rouge Hill, Port Union, Highland Creek', 'Woodbine Heights',
       'St. James Town', 'Humewood-Cedarvale',
       'Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood',
       'Guildwood, Morningside, West Hill', 'The Beaches', 'Berczy Park',
       'Caledonia-Fairbanks', 'Woburn', 'Leaside', 'Central Bay Street',
       'Christie', 'Cedarbrae', 'Hillcrest Village',
       'Bathurst Manor, Wilson Heights, Downsview North',
       'Thorncliffe Park', 'Richmond, Adelaide, King',
       'Dufferin, Dovercourt Village', 'Scarborough Village',
       'Fairview, H

In [21]:
len(df2['Neighbourhood'].unique())

99

There are no Neighbourhoods with "Not Assigned" Label

In [23]:
df2['Neighbourhood'].value_counts()[:10]

Downsview                                  4
Don Mills                                  2
Bedford Park, Lawrence Manor East          1
Davisville                                 1
Lawrence Park                              1
Birch Cliff, Cliffside West                1
Islington Avenue, Humber Valley Village    1
Parkview Hill, Woodbine Gardens            1
Humber Summit                              1
Steeles West, L'Amoreaux West              1
Name: Neighbourhood, dtype: int64

In [24]:
df2.head(3)

Postal Code           Borough              Neighbourhood
0         M3A        North York                  Parkwoods
1         M4A        North York           Victoria Village
2         M5A  Downtown Toronto  Regent Park, Harbourfront

###### Shape DataFrame

In [25]:
df2.shape

(103, 3)

There are 103 Rows in the DataFrame

---

### Part 2  - Latitudes and Longitudes

##### Import Libraries

In [26]:
import requests # library to handle requests
import pandas as pd # library for data analsysis
import numpy as np # library to handle data in a vectorized manner
import random # library for random number generation


!pip install geopy
from geopy.geocoders import Nominatim # module to convert an address into latitude and longitude values

# libraries for displaying images
from IPython.display import Image 
from IPython.core.display import HTML 
    
# tranforming json file into a pandas dataframe library
from pandas.io.json import json_normalize


! pip install folium==0.5.0
import folium # plotting library

print('Folium installed')
print('Libraries imported.')

     |████████████████████████████████| 79 kB 6.0 MB/s eta 0:00:011
  Created wheel for folium: filename=folium-0.5.0-py3-none-any.whl size=76240 sha256=3769b47e34722fb409382fa9f07e6333b89d6bd6977a665e6581b4862b04c171
  Stored in directory: /tmp/wsuser/.cache/pip/wheels/b2/2f/2c/109e446b990d663ea5ce9b078b5e7c1a9c45cca91f377080f8
Successfully built folium
Folium installed
Libraries imported.


In [27]:
! pip install geocoder

     |████████████████████████████████| 98 kB 7.1 MB/s  eta 0:00:01


In [30]:
import geocoder
g = geocoder.google('Mountain View, CA')
g.latlng

In [31]:
g

<[REQUEST_DENIED] Google - Geocode [empty]>

In [70]:
import geocoder
g = geocoder.google('Mountain View, CA')
g.latlng

In [71]:
g

<[REQUEST_DENIED] Google - Geocode [empty]>

In [72]:
g.geojson

{'type': 'FeatureCollection', 'features': []}

In [67]:
g.json

In [68]:
g.wkt

In [69]:
g.osm

In [29]:
#g = geocoder.google('Mountain View, CA')
#g.geojson
#g.json
#g.wkt
#g.osm

Apparenty, `geocoder` does not work properly at this moment. The Cells below were running for a while. I had to stop the Kernel, and provide the '#' in front of the code line, to clean up the error messages

In [28]:
#import geocoder # import geocoder
#
## initialize your variable to None
#lat_lng_coords = None
#
## loop until you get the coordinates
#while(lat_lng_coords is None):
#  g = geocoder.google('{}, Toronto, Ontario'.format(postal_code))
#  lat_lng_coords = g.latlng
#
#latitude = lat_lng_coords[0]
#longitude = lat_lng_coords[1]

In [29]:
df2.head(5)

Postal Code           Borough                                Neighbourhood
0         M3A        North York                                    Parkwoods
1         M4A        North York                             Victoria Village
2         M5A  Downtown Toronto                    Regent Park, Harbourfront
3         M6A        North York             Lawrence Manor, Lawrence Heights
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

In [30]:
df2[df2['Postal Code'] == 'M5G']

Postal Code           Borough       Neighbourhood
24         M5G  Downtown Toronto  Central Bay Street

In [31]:
#import geocoder # import geocoder
#
## initialize your variable to None
#lat_lng_coords = None
#
## loop until you get the coordinates
#while(lat_lng_coords is None):
#  g = geocoder.google('{}, Toronto, Ontario'.format('M5G'))
#  lat_lng_coords = g.latlng
#
#latitude = lat_lng_coords[0]
#longitude = lat_lng_coords[1]

#### csv file with Geographical Coordinates Toronto

In [32]:
# The code was removed by Watson Studio for sharing.

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [33]:
df2.head()

Postal Code           Borough                                Neighbourhood
0         M3A        North York                                    Parkwoods
1         M4A        North York                             Victoria Village
2         M5A  Downtown Toronto                    Regent Park, Harbourfront
3         M6A        North York             Lawrence Manor, Lawrence Heights
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

In [34]:
df_data_1.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

## Merge

In [35]:
df3 = df2.merge(df_data_1, how='left', on='Postal Code')

In [36]:
df3.head()

Postal Code           Borough                                Neighbourhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  
0  43.753259 -79.329656  
1  43.725882 -79.315572  
2  43.654260 -79.360636  
3  43.718518 -79.464763  
4  43.662301 -79.389494

In [37]:
df3.dtypes

Postal Code       object
Borough           object
Neighbourhood     object
Latitude         float64
Longitude        float64
dtype: object

In [38]:
df3.describe()

Latitude   Longitude
count  103.000000  103.000000
mean    43.704608  -79.397153
std      0.052463    0.097146
min     43.602414  -79.615819
25%     43.660567  -79.464763
50%     43.696948  -79.388790
75%     43.745320  -79.340923
max     43.836125  -79.160497

In [39]:
df3[df3['Postal Code'] == 'M5G']

Postal Code           Borough       Neighbourhood   Latitude  Longitude
24         M5G  Downtown Toronto  Central Bay Street  43.657952 -79.387383

In [40]:
df3.columns

Index(['Postal Code', 'Borough', 'Neighbourhood', 'Latitude', 'Longitude'], dtype='object')

In [41]:
df3.shape

(103, 5)

The merged file has 103 rows and 5 columns, containing ***Postal Code***, ***Borough***, ***Neighbourhood***, ***Latitude***, ***Longitude***.

___

### Part 3  - Cluster

Examine the current dataframe

In [42]:
df3.head(20)

Postal Code           Borough  \
0          M3A        North York   
1          M4A        North York   
2          M5A  Downtown Toronto   
3          M6A        North York   
4          M7A  Downtown Toronto   
5          M9A         Etobicoke   
6          M1B       Scarborough   
7          M3B        North York   
8          M4B         East York   
9          M5B  Downtown Toronto   
10         M6B        North York   
11         M9B         Etobicoke   
12         M1C       Scarborough   
13         M3C        North York   
14         M4C         East York   
15         M5C  Downtown Toronto   
16         M6C              York   
17         M9C         Etobicoke   
18         M1E       Scarborough   
19         M4E      East Toronto   

                                        Neighbourhood   Latitude  Longitude  
0                                           Parkwoods  43.753259 -79.329656  
1                                    Victoria Village  43.725882 -79.315572  
2                           Regent Park, Harbourfront  43.654260 -79.360636  
3                    Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
4         Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
5             Islington Avenue, Humber Valley Village  43.667856 -79.532242  
6                                      Malvern, Rouge  43.806686 -79.194353  
7                                           Don Mills  43.745906 -79.352188  
8                     Parkview Hill, Woodbine Gardens  43.706397 -79.309937  
9                            Garden District, Ryerson  43.657162 -79.378937  
10                                          Glencairn  43.709577 -79.445073  
11  West Deane Park, Princess Gardens, Martin Grov...  43.650943 -79.554724  
12             Rouge Hill, Port Union, Highland Creek  43.784535 -79.160497  
13                                          Don Mills  43.725900 -79.340923  
14                                   Woodbine Heights  43.695344 -79.318389  
15                                     St. James Town  43.651494 -79.375418  
16                                 Humewood-Cedarvale  43.693781 -79.428191  
17  Eringate, Bloordale Gardens, Old Burnhamthorpe...  43.643515 -79.577201  
18                  Guildwood, Morningside, West Hill  43.763573 -79.188711  
19                                        The Beaches  43.676357 -79.293031

In [43]:
print('The dataframe has {} boroughs and {} neighborhoods.'.format(
        len(df3['Borough'].unique()),
        df3.shape[0]
    )
)

The dataframe has 10 boroughs and 103 neighborhoods.


In [44]:
df3['Borough'].value_counts()

North York          24
Downtown Toronto    19
Scarborough         17
Etobicoke           12
Central Toronto      9
West Toronto         6
East York            5
York                 5
East Toronto         5
Mississauga          1
Name: Borough, dtype: int64

In [45]:
address = 'Toronto, CA'

geolocator = Nominatim(user_agent="foursquare_agent")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print(latitude, longitude)
print('The geograpical coordinates of Toronto are {}, {}.'.format(latitude, longitude))

43.6534817 -79.3839347
The geograpical coordinates of Toronto are 43.6534817, -79.3839347.


### Map of Greater Toronto with neighborhoods superimposed on top.¶

In [47]:
# create map of Greater Toronto using latitude and longitude values
map_toronto_great = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(df3['Latitude'], df3['Longitude'], df3['Borough'], df3['Neighbourhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto_great)  
    
map_toronto_great

#### Toronto

Work with only Boroughs that contain the word Toronto 

In [48]:
df3['Borough'].unique()

array(['North York', 'Downtown Toronto', 'Etobicoke', 'Scarborough',
       'East York', 'York', 'East Toronto', 'West Toronto',
       'Central Toronto', 'Mississauga'], dtype=object)

In [49]:
df3[df3['Borough'] == 'East Toronto']

Postal Code       Borough  \
19          M4E  East Toronto   
41          M4K  East Toronto   
47          M4L  East Toronto   
54          M4M  East Toronto   
100         M7Y  East Toronto   

                                         Neighbourhood   Latitude  Longitude  
19                                         The Beaches  43.676357 -79.293031  
41                        The Danforth West, Riverdale  43.679557 -79.352188  
47                      India Bazaar, The Beaches West  43.668999 -79.315572  
54                                     Studio District  43.659526 -79.340923  
100  Business reply mail Processing Centre, South C...  43.662744 -79.321558

In [50]:
df4 = df3.loc[df3['Borough'].str.endswith('Toronto')]

In [51]:
df4.shape

(39, 5)

In [52]:
df4['Borough'].value_counts()

Downtown Toronto    19
Central Toronto      9
West Toronto         6
East Toronto         5
Name: Borough, dtype: int64

In [53]:
print('The new Toronto dataframe has {} boroughs and {} neighborhoods.'.format(
        len(df4['Borough'].unique()),
        df4.shape[0]
    )
)

The new Toronto dataframe has 4 boroughs and 39 neighborhoods.


In [54]:
df4.columns

Index(['Postal Code', 'Borough', 'Neighbourhood', 'Latitude', 'Longitude'], dtype='object')

In [55]:
df4['Borough'].unique()

array(['Downtown Toronto', 'East Toronto', 'West Toronto',
       'Central Toronto'], dtype=object)

The only Boroughs that contain the word Toronto are *'Downtown Toronto'*, *'East Toronto'*, *'West Toronto'* and *'Central Toronto'*.

In [56]:
df4

Postal Code           Borough  \
2           M5A  Downtown Toronto   
4           M7A  Downtown Toronto   
9           M5B  Downtown Toronto   
15          M5C  Downtown Toronto   
19          M4E      East Toronto   
20          M5E  Downtown Toronto   
24          M5G  Downtown Toronto   
25          M6G  Downtown Toronto   
30          M5H  Downtown Toronto   
31          M6H      West Toronto   
36          M5J  Downtown Toronto   
37          M6J      West Toronto   
41          M4K      East Toronto   
42          M5K  Downtown Toronto   
43          M6K      West Toronto   
47          M4L      East Toronto   
48          M5L  Downtown Toronto   
54          M4M      East Toronto   
61          M4N   Central Toronto   
62          M5N   Central Toronto   
67          M4P   Central Toronto   
68          M5P   Central Toronto   
69          M6P      West Toronto   
73          M4R   Central Toronto   
74          M5R   Central Toronto   
75          M6R      West Toronto   
79          M4S   Central Toronto   
80          M5S  Downtown Toronto   
81          M6S      West Toronto   
83          M4T   Central Toronto   
84          M5T  Downtown Toronto   
86          M4V   Central Toronto   
87          M5V  Downtown Toronto   
91          M4W  Downtown Toronto   
92          M5W  Downtown Toronto   
96          M4X  Downtown Toronto   
97          M5X  Downtown Toronto   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   

                                         Neighbourhood   Latitude  Longitude  
2                            Regent Park, Harbourfront  43.654260 -79.360636  
4          Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
9                             Garden District, Ryerson  43.657162 -79.378937  
15                                      St. James Town  43.651494 -79.375418  
19                                         The Beaches  43.676357 -79.293031  
20                                         Berczy Park  43.644771 -79.373306  
24                                  Central Bay Street  43.657952 -79.387383  
25                                            Christie  43.669542 -79.422564  
30                            Richmond, Adelaide, King  43.650571 -79.384568  
31                        Dufferin, Dovercourt Village  43.669005 -79.442259  
36   Harbourfront East, Union Station, Toronto Islands  43.640816 -79.381752  
37                            Little Portugal, Trinity  43.647927 -79.419750  
41                        The Danforth West, Riverdale  43.679557 -79.352188  
42            Toronto Dominion Centre, Design Exchange  43.647177 -79.381576  
43        Brockton, Parkdale Village, Exhibition Place  43.636847 -79.428191  
47                      India Bazaar, The Beaches West  43.668999 -79.315572  
48                      Commerce Court, Victoria Hotel  43.648198 -79.379817  
54                                     Studio District  43.659526 -79.340923  
61                                       Lawrence Park  43.728020 -79.388790  
62                                            Roselawn  43.711695 -79.416936  
67                                    Davisville North  43.712751 -79.390197  
68     Forest Hill North & West, Forest Hill Road Park  43.696948 -79.411307  
69                       High Park, The Junction South  43.661608 -79.464763  
73                   North Toronto West, Lawrence Park  43.715383 -79.405678  
74                 The Annex, North Midtown, Yorkville  43.672710 -79.405678  
75                              Parkdale, Roncesvalles  43.648960 -79.456325  
79                                          Davisville  43.704324 -79.388790  
80                      University of Toronto, Harbord  43.662696 -79.400049  
81                                  Runnymede, Swansea  43.651571 -79.484450  
83                         Moore Park, Summerhill East  43.689574 -79.383160  
84           Kensington Market, Chinatown, Grange Park  43.653206 -79.4000

In [57]:
df4.head()

Postal Code           Borough                                Neighbourhood  \
2          M5A  Downtown Toronto                    Regent Park, Harbourfront   
4          M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
9          M5B  Downtown Toronto                     Garden District, Ryerson   
15         M5C  Downtown Toronto                               St. James Town   
19         M4E      East Toronto                                  The Beaches   

     Latitude  Longitude  
2   43.654260 -79.360636  
4   43.662301 -79.389494  
9   43.657162 -79.378937  
15  43.651494 -79.375418  
19  43.676357 -79.293031

###### Reset Index

In [58]:
df4.reset_index(inplace=True)

In [59]:
df4.head(3)

index Postal Code           Borough  \
0      2         M5A  Downtown Toronto   
1      4         M7A  Downtown Toronto   
2      9         M5B  Downtown Toronto   

                                 Neighbourhood   Latitude  Longitude  
0                    Regent Park, Harbourfront  43.654260 -79.360636  
1  Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
2                     Garden District, Ryerson  43.657162 -79.378937

In [60]:
df4.drop('index', axis=1, inplace=True)

/opt/conda/envs/Python-3.7-main/lib/python3.7/site-packages/pandas/core/frame.py:3997: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  errors=errors,


In [61]:
df4.head(3)

Postal Code           Borough                                Neighbourhood  \
0         M5A  Downtown Toronto                    Regent Park, Harbourfront   
1         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
2         M5B  Downtown Toronto                     Garden District, Ryerson   

    Latitude  Longitude  
0  43.654260 -79.360636  
1  43.662301 -79.389494  
2  43.657162 -79.378937

In [62]:
df4

Postal Code           Borough  \
0          M5A  Downtown Toronto   
1          M7A  Downtown Toronto   
2          M5B  Downtown Toronto   
3          M5C  Downtown Toronto   
4          M4E      East Toronto   
5          M5E  Downtown Toronto   
6          M5G  Downtown Toronto   
7          M6G  Downtown Toronto   
8          M5H  Downtown Toronto   
9          M6H      West Toronto   
10         M5J  Downtown Toronto   
11         M6J      West Toronto   
12         M4K      East Toronto   
13         M5K  Downtown Toronto   
14         M6K      West Toronto   
15         M4L      East Toronto   
16         M5L  Downtown Toronto   
17         M4M      East Toronto   
18         M4N   Central Toronto   
19         M5N   Central Toronto   
20         M4P   Central Toronto   
21         M5P   Central Toronto   
22         M6P      West Toronto   
23         M4R   Central Toronto   
24         M5R   Central Toronto   
25         M6R      West Toronto   
26         M4S   Central Toronto   
27         M5S  Downtown Toronto   
28         M6S      West Toronto   
29         M4T   Central Toronto   
30         M5T  Downtown Toronto   
31         M4V   Central Toronto   
32         M5V  Downtown Toronto   
33         M4W  Downtown Toronto   
34         M5W  Downtown Toronto   
35         M4X  Downtown Toronto   
36         M5X  Downtown Toronto   
37         M4Y  Downtown Toronto   
38         M7Y      East Toronto   

                                        Neighbourhood   Latitude  Longitude  
0                           Regent Park, Harbourfront  43.654260 -79.360636  
1         Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
2                            Garden District, Ryerson  43.657162 -79.378937  
3                                      St. James Town  43.651494 -79.375418  
4                                         The Beaches  43.676357 -79.293031  
5                                         Berczy Park  43.644771 -79.373306  
6                                  Central Bay Street  43.657952 -79.387383  
7                                            Christie  43.669542 -79.422564  
8                            Richmond, Adelaide, King  43.650571 -79.384568  
9                        Dufferin, Dovercourt Village  43.669005 -79.442259  
10  Harbourfront East, Union Station, Toronto Islands  43.640816 -79.381752  
11                           Little Portugal, Trinity  43.647927 -79.419750  
12                       The Danforth West, Riverdale  43.679557 -79.352188  
13           Toronto Dominion Centre, Design Exchange  43.647177 -79.381576  
14       Brockton, Parkdale Village, Exhibition Place  43.636847 -79.428191  
15                     India Bazaar, The Beaches West  43.668999 -79.315572  
16                     Commerce Court, Victoria Hotel  43.648198 -79.379817  
17                                    Studio District  43.659526 -79.340923  
18                                      Lawrence Park  43.728020 -79.388790  
19                                           Roselawn  43.711695 -79.416936  
20                                   Davisville North  43.712751 -79.390197  
21    Forest Hill North & West, Forest Hill Road Park  43.696948 -79.411307  
22                      High Park, The Junction South  43.661608 -79.464763  
23                  North Toronto West, Lawrence Park  43.715383 -79.405678  
24                The Annex, North Midtown, Yorkville  43.672710 -79.405678  
25                             Parkdale, Roncesvalles  43.648960 -79.456325  
26                                         Davisville  43.704324 -79.388790  
27                     University of Toronto, Harbord  43.662696 -79.400049  
28                                 Runnymede, Swansea  43.651571 -79.484450  
29                        Moore Park, Summerhill East  43.689574 -79.383160  
30          Kensington Market, Chinatown, Grange Park  43.653206 -79.400049  
31  Summerhill West, Rathnelly, South Hill, Forest...  43.686412 -

In [63]:
# create map of Toronto (City) using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(df4['Latitude'], df4['Longitude'], df4['Borough'], df4['Neighbourhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

### Foursquare API to explore the neighborhoods and segment them.

##### Define Foursquare Credentials and Version¶

In [64]:
# The code was removed by Watson Studio for sharing.

### 1. First neighborhood in Toronto dataframe.

In [65]:
df4.loc[0, 'Neighbourhood']

'Regent Park, Harbourfront'

In [66]:
neighborhood_latitude = df4.loc[0, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = df4.loc[0, 'Longitude'] # neighborhood longitude value

neighborhood_name = df4.loc[0, 'Neighbourhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of Regent Park, Harbourfront are 43.6542599, -79.3606359.


In [67]:
radius = 500

In [68]:
# The code was removed by Watson Studio for sharing.

In [69]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5fa8fa693a47d452b8d70dea'},
 'response': {'headerLocation': 'Corktown',
  'headerFullLocation': 'Corktown, Toronto',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 44,
  'suggestedBounds': {'ne': {'lat': 43.6587599045, 'lng': -79.3544279001486},
   'sw': {'lat': 43.6497598955, 'lng': -79.36684389985142}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '54ea41ad498e9a11e9e13308',
       'name': 'Roselle Desserts',
       'location': {'address': '362 King St E',
        'crossStreet': 'Trinity St',
        'lat': 43.653446723052674,
        'lng': -79.3620167174383,
        'labeledLatLngs': [{'label': 'display',
          'lat': 43.653446723052674,
          'lng': -79.3620167174383}],
        'distance': 143,
       

In [70]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [71]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head(44)

/opt/conda/envs/Python-3.7-main/lib/python3.7/site-packages/ipykernel/__main__.py:3: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead
  app.launch_new_instance()


name             categories        lat  \
0                       Roselle Desserts                 Bakery  43.653447   
1                          Tandem Coffee            Coffee Shop  43.653559   
2                 Cooper Koo Family YMCA    Distribution Center  43.653249   
3                         Impact Kitchen             Restaurant  43.656369   
4                    Body Blitz Spa East                    Spa  43.654735   
5                        Corktown Common                   Park  43.655618   
6               Dominion Pub and Kitchen                    Pub  43.656919   
7                 Figs Breakfast & Lunch         Breakfast Spot  43.655675   
8                     The Extension Room   Gym / Fitness Center  43.653313   
9       The Distillery Historic District          Historic Site  43.650244   
10                    Morning Glory Cafe         Breakfast Spot  43.653947   
11                   Brick Street Bakery                 Bakery  43.650574   
12              Distillery Sunday Market         Farmers Market  43.650075   
13                   SOMA chocolatemaker         Chocolate Shop  43.650622   
14                        Rooster Coffee            Coffee Shop  43.651900   
15                       Sumach Espresso            Coffee Shop  43.658135   
16                                  Arvo            Coffee Shop  43.649963   
17               Dark Horse Espresso Bar            Coffee Shop  43.653081   
18                             Starbucks            Coffee Shop  43.651613   
19  Young Centre for the Performing Arts  Performing Arts Venue  43.650825   
20                        Underpass Park                   Park  43.655764   
21                       Alumnae Theatre                Theater  43.652756   
22            Cluny Bistro & Boulangerie      French Restaurant  43.650565   
23                              Cacao 70           Dessert Shop  43.650067   
24                       The Yoga Lounge            Yoga Studio  43.655515   
25                Parliament Square Park                   Park  43.650264   
26           The Sweet Escape Patisserie                 Bakery  43.650632   
27                           Caffe Furbo                   Café  43.649970   
28                             El Catrin     Mexican Restaurant  43.650601   
29                     Mill St. Brew Pub                    Pub  43.650353   
30                    Soulpepper Theatre                Theater  43.650780   
31                       Berkeley Church            Event Space  43.655123   
32                    John Fluevog Shoes             Shoe Store  43.649896   
33                       ODIN Cafe + Bar                   Café  43.656739   
34                            The Aviary                    Pub  43.653634   
35                          Arta Gallery            Art Gallery  43.650022   
36                                Vistek      Electronics Store  43.657046   
37     Ontario Spring Water Sake Company                Brewery  43.649922   
38                       TD Canada Trust                   Bank  43.652780   
39                        The Beer Store             Beer Store  43.657773   
40                       Savoury Grounds            Coffee Shop  43.656821   
41         Residence & Conference Centre                  Hotel  43.653040   
42                                 FUEL+            Coffee Shop  43.653193   
43                            GW General           Antique Shop  43.650495   

          lng  
0  -79.362017  
1  -79.361809  
2  -79.358008  
3  -79.356980  
4  -79.359874  
5  -79.356211  
6  -79.358967  
7  -79.364503  
8  -79.359725  
9  -79.359323  
10 -79.361149  
11 -79.359539  
12 -79.361832  
13 -79.358127  
14 -79.365609  
15 -79.359515  
16 -79.361442  
17 -79.357078  
18 -79.364917  
19 -79.357593  
20 -79.354806  
21 -79.364753  
22 -79.357843  
23 -79.360723  
24 -79.364955  
25 -79.362195  
26 -79.358709  
27 -79.358849  
28 -79.358920  
29 -79.358489  
30 -79.357615  
31 -79.365873

In [72]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

44 venues were returned by Foursquare.


### 2. Explore All Neighborhoods in Toronto City Dataframe.

In [73]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [74]:
lat

43.6627439

In [75]:
names = df4.loc[:, 'Neighbourhood']

In [76]:
names[:10]

0                      Regent Park, Harbourfront
1    Queen's Park, Ontario Provincial Government
2                       Garden District, Ryerson
3                                 St. James Town
4                                    The Beaches
5                                    Berczy Park
6                             Central Bay Street
7                                       Christie
8                       Richmond, Adelaide, King
9                   Dufferin, Dovercourt Village
Name: Neighbourhood, dtype: object

In [77]:
latitudes = df4.loc[:, 'Latitude'] # neighborhood latitude value

In [78]:
longitudes = df4.loc[:, 'Longitude']

In [79]:
getNearbyVenues(names, latitudes, longitudes, radius=500)

Regent Park, Harbourfront
Queen's Park, Ontario Provincial Government
Garden District, Ryerson
St. James Town
The Beaches
Berczy Park
Central Bay Street
Christie
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
The Danforth West, Riverdale
Toronto Dominion Centre, Design Exchange
Brockton, Parkdale Village, Exhibition Place
India Bazaar, The Beaches West
Commerce Court, Victoria Hotel
Studio District
Lawrence Park
Roselawn
Davisville North
Forest Hill North & West, Forest Hill Road Park
High Park, The Junction South
North Toronto West, Lawrence Park
The Annex, North Midtown, Yorkville
Parkdale, Roncesvalles
Davisville
University of Toronto, Harbord
Runnymede, Swansea
Moore Park, Summerhill East
Kensington Market, Chinatown, Grange Park
Summerhill West, Rathnelly, South Hill, Forest Hill SE, Deer Park
CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport
R

Neighborhood  \
0                             Regent Park, Harbourfront   
1                             Regent Park, Harbourfront   
2                             Regent Park, Harbourfront   
3                             Regent Park, Harbourfront   
4                             Regent Park, Harbourfront   
...                                                 ...   
1619  Business reply mail Processing Centre, South C...   
1620  Business reply mail Processing Centre, South C...   
1621  Business reply mail Processing Centre, South C...   
1622  Business reply mail Processing Centre, South C...   
1623  Business reply mail Processing Centre, South C...   

      Neighborhood Latitude  Neighborhood Longitude                    Venue  \
0                 43.654260              -79.360636         Roselle Desserts   
1                 43.654260              -79.360636            Tandem Coffee   
2                 43.654260              -79.360636   Cooper Koo Family YMCA   
3                 43.654260              -79.360636           Impact Kitchen   
4                 43.654260              -79.360636      Body Blitz Spa East   
...                     ...                     ...                      ...   
1619              43.662744              -79.321558     TTC Russell Division   
1620              43.662744              -79.321558  Jonathan Ashbridge Park   
1621              43.662744              -79.321558         Olliffe On Queen   
1622              43.662744              -79.321558              ONE Academy   
1623              43.662744              -79.321558     Revolution Recording   

      Venue Latitude  Venue Longitude        Venue Category  
0          43.653447       -79.362017                Bakery  
1          43.653559       -79.361809           Coffee Shop  
2          43.653249       -79.358008   Distribution Center  
3          43.656369       -79.356980            Restaurant  
4          43.654735       -79.359874                   Spa  
...              ...              ...                   ...  
1619       43.664908       -79.322560    Light Rail Station  
1620       43.664702       -79.319898                  Park  
1621       43.664503       -79.324768               Butcher  
1622       43.662253       -79.326911  Gym / Fitness Center  
1623       43.662561       -79.326940      Recording Studio  

[1624 rows x 7 columns]

In [82]:
nearby_venues[0:5]

name           categories        lat        lng
0        Roselle Desserts               Bakery  43.653447 -79.362017
1           Tandem Coffee          Coffee Shop  43.653559 -79.361809
2  Cooper Koo Family YMCA  Distribution Center  43.653249 -79.358008
3          Impact Kitchen           Restaurant  43.656369 -79.356980
4     Body Blitz Spa East                  Spa  43.654735 -79.359874

In [85]:
getNearbyVenues(names, latitudes, longitudes, radius=500)

Regent Park, Harbourfront
Queen's Park, Ontario Provincial Government
Garden District, Ryerson
St. James Town
The Beaches
Berczy Park
Central Bay Street
Christie
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
The Danforth West, Riverdale
Toronto Dominion Centre, Design Exchange
Brockton, Parkdale Village, Exhibition Place
India Bazaar, The Beaches West
Commerce Court, Victoria Hotel
Studio District
Lawrence Park
Roselawn
Davisville North
Forest Hill North & West, Forest Hill Road Park
High Park, The Junction South
North Toronto West, Lawrence Park
The Annex, North Midtown, Yorkville
Parkdale, Roncesvalles
Davisville
University of Toronto, Harbord
Runnymede, Swansea
Moore Park, Summerhill East
Kensington Market, Chinatown, Grange Park
Summerhill West, Rathnelly, South Hill, Forest Hill SE, Deer Park
CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport
R

Neighborhood  \
0                             Regent Park, Harbourfront   
1                             Regent Park, Harbourfront   
2                             Regent Park, Harbourfront   
3                             Regent Park, Harbourfront   
4                             Regent Park, Harbourfront   
...                                                 ...   
1619  Business reply mail Processing Centre, South C...   
1620  Business reply mail Processing Centre, South C...   
1621  Business reply mail Processing Centre, South C...   
1622  Business reply mail Processing Centre, South C...   
1623  Business reply mail Processing Centre, South C...   

      Neighborhood Latitude  Neighborhood Longitude                    Venue  \
0                 43.654260              -79.360636         Roselle Desserts   
1                 43.654260              -79.360636            Tandem Coffee   
2                 43.654260              -79.360636   Cooper Koo Family YMCA   
3                 43.654260              -79.360636           Impact Kitchen   
4                 43.654260              -79.360636      Body Blitz Spa East   
...                     ...                     ...                      ...   
1619              43.662744              -79.321558     TTC Russell Division   
1620              43.662744              -79.321558  Jonathan Ashbridge Park   
1621              43.662744              -79.321558         Olliffe On Queen   
1622              43.662744              -79.321558              ONE Academy   
1623              43.662744              -79.321558     Revolution Recording   

      Venue Latitude  Venue Longitude        Venue Category  
0          43.653447       -79.362017                Bakery  
1          43.653559       -79.361809           Coffee Shop  
2          43.653249       -79.358008   Distribution Center  
3          43.656369       -79.356980            Restaurant  
4          43.654735       -79.359874                   Spa  
...              ...              ...                   ...  
1619       43.664908       -79.322560    Light Rail Station  
1620       43.664702       -79.319898                  Park  
1621       43.664503       -79.324768               Butcher  
1622       43.662253       -79.326911  Gym / Fitness Center  
1623       43.662561       -79.326940      Recording Studio  

[1624 rows x 7 columns]

In [86]:
toronto_venues = getNearbyVenues(names, latitudes, longitudes, radius=500)

Regent Park, Harbourfront
Queen's Park, Ontario Provincial Government
Garden District, Ryerson
St. James Town
The Beaches
Berczy Park
Central Bay Street
Christie
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
The Danforth West, Riverdale
Toronto Dominion Centre, Design Exchange
Brockton, Parkdale Village, Exhibition Place
India Bazaar, The Beaches West
Commerce Court, Victoria Hotel
Studio District
Lawrence Park
Roselawn
Davisville North
Forest Hill North & West, Forest Hill Road Park
High Park, The Junction South
North Toronto West, Lawrence Park
The Annex, North Midtown, Yorkville
Parkdale, Roncesvalles
Davisville
University of Toronto, Harbord
Runnymede, Swansea
Moore Park, Summerhill East
Kensington Market, Chinatown, Grange Park
Summerhill West, Rathnelly, South Hill, Forest Hill SE, Deer Park
CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport
R

In [87]:
print(toronto_venues.shape)

(1624, 7)


In [88]:
toronto_venues.head()

Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  Regent Park, Harbourfront               43.65426              -79.360636   
1  Regent Park, Harbourfront               43.65426              -79.360636   
2  Regent Park, Harbourfront               43.65426              -79.360636   
3  Regent Park, Harbourfront               43.65426              -79.360636   
4  Regent Park, Harbourfront               43.65426              -79.360636   

                    Venue  Venue Latitude  Venue Longitude  \
0        Roselle Desserts       43.653447       -79.362017   
1           Tandem Coffee       43.653559       -79.361809   
2  Cooper Koo Family YMCA       43.653249       -79.358008   
3          Impact Kitchen       43.656369       -79.356980   
4     Body Blitz Spa East       43.654735       -79.359874   

        Venue Category  
0               Bakery  
1          Coffee Shop  
2  Distribution Center  
3           Restaurant  
4                  Spa

In [89]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Berczy Park                                                            55   
Brockton, Parkdale Village, Exhibition Place                           23   
Business reply mail Processing Centre, South Ce...                     16   
CN Tower, King and Spadina, Railway Lands, Harb...                     16   
Central Bay Street                                                     68   
Christie                                                               16   
Church and Wellesley                                                   75   
Commerce Court, Victoria Hotel                                        100   
Davisville                                                             33   
Davisville North                                                        9   
Dufferin, Dovercourt Village                                           13   
First Canadian Place, Underground city                                100   
Forest Hill North & West, Forest Hill Road Park                         4   
Garden District, Ryerson                                              100   
Harbourfront East, Union Station, Toronto Islands                     100   
High Park, The Junction South                                          25   
India Bazaar, The Beaches West                                         19   
Kensington Market, Chinatown, Grange Park                              74   
Lawrence Park                                                           3   
Little Portugal, Trinity                                               45   
Moore Park, Summerhill East                                             2   
North Toronto West, Lawrence Park                                      18   
Parkdale, Roncesvalles                                                 14   
Queen's Park, Ontario Provincial Government                            33   
Regent Park, Harbourfront                                              44   
Richmond, Adelaide, King                                              100   
Rosedale                                                                4   
Roselawn                                                                2   
Runnymede, Swansea                                                     33   
St. James Town                                                         85   
St. James Town, Cabbagetown                                            48   
Stn A PO Boxes                                                         96   
Studio District                                                        37   
Summerhill West, Rathnelly, South Hill, Forest ...                     14   
The Annex, North Midtown, Yorkville                                    19   
The Beaches                                                             4   
The Danforth West, Riverdale                                           43   
Toronto Dominion Centre, Design Exchange                              100   
University of Toronto, Harbord                                         34   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Berczy Park                                                             55   
Brockton, Parkdale Village, Exhibition Place                            23   
Business reply mail Processing Centre, South Ce...                      16   
CN Tower, King and Spadina, Railway Lands, Harb...                      16   
Central Bay Street                                                      68   
Christie                                                                16   
Church and Wellesley                                                    75   
Commerce Court, Victoria Hotel                                         100   
Davisville                                                              33   
Davisville North                    

In [90]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 236 uniques categories.


### 3. Analyze Neighborhoods Toronto

In [91]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Yoga Studio  Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0            0                  0        0                   0             0   
1            0                  0        0                   0             0   
2            0                  0        0                   0             0   
3            0                  0        0                   0             0   
4            0                  0        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  ...  Theater  Theme Restaurant  Toy / Game Store  Trail  \
0             0  ...        0                 0                 0      0   
1             0  ...        0                 0                 0      0   
2             0  ...        0                 0                 0      0   
3             0  ...        0                 0                 0      0   
4             0  ...        0                 0                 0      0   

   Train Station  Vegetarian / Vegan Restaurant  Video Game Store  \
0              0                              0                 0   
1              0                              0                 0   
2              0                              0                 0   
3              0                              0                 0   
4              0                              0                 0   

   Vietnamese Restaurant  Wine Bar  Women's Store  
0                      0         0              0  
1                      0         0              0  
2                      0         0              0  
3                      0         0              0  
4                      0         0              0  

[5 rows x 236 columns]

In [92]:
toronto_onehot.shape

(1624, 236)

##### Grouped rows by neighborhood and Mean of  Frequency of Occurrence of each category

In [93]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()

In [94]:
toronto_grouped

Neighborhood  Yoga Studio  \
0                                         Berczy Park     0.000000   
1        Brockton, Parkdale Village, Exhibition Place     0.000000   
2   Business reply mail Processing Centre, South C...     0.000000   
3   CN Tower, King and Spadina, Railway Lands, Har...     0.000000   
4                                  Central Bay Street     0.014706   
5                                            Christie     0.000000   
6                                Church and Wellesley     0.026667   
7                      Commerce Court, Victoria Hotel     0.000000   
8                                          Davisville     0.000000   
9                                    Davisville North     0.000000   
10                       Dufferin, Dovercourt Village     0.000000   
11             First Canadian Place, Underground city     0.000000   
12    Forest Hill North & West, Forest Hill Road Park     0.000000   
13                           Garden District, Ryerson     0.000000   
14  Harbourfront East, Union Station, Toronto Islands     0.000000   
15                      High Park, The Junction South     0.000000   
16                     India Bazaar, The Beaches West     0.000000   
17          Kensington Market, Chinatown, Grange Park     0.000000   
18                                      Lawrence Park     0.000000   
19                           Little Portugal, Trinity     0.022222   
20                        Moore Park, Summerhill East     0.000000   
21                  North Toronto West, Lawrence Park     0.055556   
22                             Parkdale, Roncesvalles     0.000000   
23        Queen's Park, Ontario Provincial Government     0.030303   
24                          Regent Park, Harbourfront     0.022727   
25                           Richmond, Adelaide, King     0.000000   
26                                           Rosedale     0.000000   
27                                           Roselawn     0.000000   
28                                 Runnymede, Swansea     0.030303   
29                                     St. James Town     0.000000   
30                        St. James Town, Cabbagetown     0.000000   
31                                     Stn A PO Boxes     0.010417   
32                                    Studio District     0.027027   
33  Summerhill West, Rathnelly, South Hill, Forest...     0.000000   
34                The Annex, North Midtown, Yorkville     0.000000   
35                                        The Beaches     0.000000   
36                       The Danforth West, Riverdale     0.023256   
37           Toronto Dominion Centre, Design Exchange     0.000000   
38                     University of Toronto, Harbord     0.029412   

    Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0            0.000000   0.0000              0.0000        0.0000   
1            0.000000   0.0000              0.0000        0.0000   
2            0.000000   0.0000              0.0000        0.0000   
3            0.000000   0.0625              0.0625        0.0625   
4            0.000000   0.0000              0.0000        0.0000   
5            0.000000   0.0000              0.0000        0.0000   
6            0.013333   0.0000              0.0000        0.0000   
7            0.000000   0.0000              0.0000        0.0000   
8            0.000000   0.0000              0.0000        0.0000   
9            0.000000   0.0000              0.0000        0.0000   
10           0.000000   0.0000              0.0000        0.0000   
11           0.000000   0.0000              0.0000        0.0000   
12           0.000000   0.0000              0.0000        0.0000   
13           0.000000   0.0000              0.0000        0.0000   
14           0.000000   0.0000              0.0000        0.0000   
15           0.000000   0.0000              0.0000        0.0000   
16           0.000000   0.0000              0.0000        0.0000   
17           0.0

In [95]:
toronto_grouped.head(10)

Neighborhood  Yoga Studio  \
0                                        Berczy Park     0.000000   
1       Brockton, Parkdale Village, Exhibition Place     0.000000   
2  Business reply mail Processing Centre, South C...     0.000000   
3  CN Tower, King and Spadina, Railway Lands, Har...     0.000000   
4                                 Central Bay Street     0.014706   
5                                           Christie     0.000000   
6                               Church and Wellesley     0.026667   
7                     Commerce Court, Victoria Hotel     0.000000   
8                                         Davisville     0.000000   
9                                   Davisville North     0.000000   

   Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0           0.000000   0.0000              0.0000        0.0000   
1           0.000000   0.0000              0.0000        0.0000   
2           0.000000   0.0000              0.0000        0.0000   
3           0.000000   0.0625              0.0625        0.0625   
4           0.000000   0.0000              0.0000        0.0000   
5           0.000000   0.0000              0.0000        0.0000   
6           0.013333   0.0000              0.0000        0.0000   
7           0.000000   0.0000              0.0000        0.0000   
8           0.000000   0.0000              0.0000        0.0000   
9           0.000000   0.0000              0.0000        0.0000   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0           0.000            0.000            0.0000             0.000000   
1           0.000            0.000            0.0000             0.000000   
2           0.000            0.000            0.0000             0.000000   
3           0.125            0.125            0.0625             0.000000   
4           0.000            0.000            0.0000             0.000000   
5           0.000            0.000            0.0000             0.000000   
6           0.000            0.000            0.0000             0.013333   
7           0.000            0.000            0.0000             0.040000   
8           0.000            0.000            0.0000             0.000000   
9           0.000            0.000            0.0000             0.000000   

   ...   Theater  Theme Restaurant  Toy / Game Store  Trail  Train Station  \
0  ...  0.000000          0.000000          0.000000    0.0            0.0   
1  ...  0.000000          0.000000          0.000000    0.0            0.0   
2  ...  0.000000          0.000000          0.000000    0.0            0.0   
3  ...  0.000000          0.000000          0.000000    0.0            0.0   
4  ...  0.000000          0.000000          0.000000    0.0            0.0   
5  ...  0.000000          0.000000          0.000000    0.0            0.0   
6  ...  0.013333          0.013333          0.000000    0.0            0.0   
7  ...  0.000000          0.000000          0.000000    0.0            0.0   
8  ...  0.000000          0.000000          0.030303    0.0            0.0   
9  ...  0.000000          0.000000          0.000000    0.0            0.0   

   Vegetarian / Vegan Restaurant  Video Game Store  Vietnamese Restaurant  \
0                       0.018182               0.0               0.000000   
1                       0.000000               0.0               0.000000   
2                       0.000000               0.0               0.000000   
3                       0.000000               0.0               0.000000   
4                       0.014706               0.0               0.000000   
5                       0.000000               0.0               0.000000   
6                       0.000000               0.0               0.013333   
7                       0.020000               0.0               0.000000   
8                       0.000000               0.0               0.000000   
9                       0.000000               0.0               

In [96]:
toronto_grouped.shape

(39, 236)

##### Neighborhood with the top 5 most common venues

In [97]:
num_top_venues = 5

for hood in toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Berczy Park----
                venue  freq
0         Coffee Shop  0.09
1  Seafood Restaurant  0.04
2         Cheese Shop  0.04
3            Beer Bar  0.04
4        Cocktail Bar  0.04


----Brockton, Parkdale Village, Exhibition Place----
            venue  freq
0            Café  0.13
1       Nightclub  0.09
2     Coffee Shop  0.09
3  Breakfast Spot  0.09
4   Grocery Store  0.04


----Business reply mail Processing Centre, South Central Letter Processing Plant Toronto----
                  venue  freq
0  Fast Food Restaurant  0.06
1         Auto Workshop  0.06
2            Comic Shop  0.06
3           Pizza Place  0.06
4      Recording Studio  0.06


----CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport----
             venue  freq
0   Airport Lounge  0.12
1  Airport Service  0.12
2  Harbor / Marina  0.06
3            Plane  0.06
4    Boat or Ferry  0.06


----Central Bay Street----
                venue  freq
0         Coff

#### Top Venues in Dataframe

##### Function to sort the venues in descending order

In [98]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

##### New Dataframe with Top 10 Venues for each Neighborhood

In [123]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Berczy Park           Coffee Shop   
1       Brockton, Parkdale Village, Exhibition Place                  Café   
2  Business reply mail Processing Centre, South C...            Skate Park   
3  CN Tower, King and Spadina, Railway Lands, Har...        Airport Lounge   
4                                 Central Bay Street           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0        Farmers Market                Bakery          Cocktail Bar   
1           Coffee Shop        Breakfast Spot             Nightclub   
2           Pizza Place               Brewery         Burrito Place   
3       Airport Service           Coffee Shop       Harbor / Marina   
4                  Café        Sandwich Place    Italian Restaurant   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0    Seafood Restaurant           Cheese Shop              Beer Bar   
1             Pet Store                   Bar         Burrito Place   
2            Restaurant        Farmers Market  Fast Food Restaurant   
3                 Plane   Rental Car Location      Sculpture Garden   
4           Salad Place          Burger Joint      Department Store   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0            Restaurant         Grocery Store                    Pub  
1            Restaurant          Climbing Gym  Performing Arts Venue  
2      Recording Studio               Butcher     Light Rail Station  
3              Boutique                   Bar          Boat or Ferry  
4   Japanese Restaurant       Thai Restaurant        Bubble Tea Shop

### 4. Cluster Neighborhoods

k-means to cluster the neighborhood into 5 clusters.

In [124]:
# import k-means from clustering stage
from sklearn.cluster import KMeans

In [125]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0], dtype=int32)

New dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [126]:
toronto_data = df4

In [104]:
df4.head()

Postal Code           Borough                                Neighbourhood  \
0         M5A  Downtown Toronto                    Regent Park, Harbourfront   
1         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
2         M5B  Downtown Toronto                     Garden District, Ryerson   
3         M5C  Downtown Toronto                               St. James Town   
4         M4E      East Toronto                                  The Beaches   

    Latitude  Longitude  
0  43.654260 -79.360636  
1  43.662301 -79.389494  
2  43.657162 -79.378937  
3  43.651494 -79.375418  
4  43.676357 -79.293031

In [116]:
toronto_data.head()

Postal Code           Borough                                Neighbourhood  \
0         M5A  Downtown Toronto                    Regent Park, Harbourfront   
1         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
2         M5B  Downtown Toronto                     Garden District, Ryerson   
3         M5C  Downtown Toronto                               St. James Town   
4         M4E      East Toronto                                  The Beaches   

    Latitude  Longitude  
0  43.654260 -79.360636  
1  43.662301 -79.389494  
2  43.657162 -79.378937  
3  43.651494 -79.375418  
4  43.676357 -79.293031

In [117]:
toronto_data['Neighborhood'] = toronto_data['Neighbourhood']

/opt/conda/envs/Python-3.7-main/lib/python3.7/site-packages/ipykernel/__main__.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  if __name__ == '__main__':


In [118]:
toronto_data.head()

Postal Code           Borough                                Neighbourhood  \
0         M5A  Downtown Toronto                    Regent Park, Harbourfront   
1         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
2         M5B  Downtown Toronto                     Garden District, Ryerson   
3         M5C  Downtown Toronto                               St. James Town   
4         M4E      East Toronto                                  The Beaches   

    Latitude  Longitude                                 Neighborhood  
0  43.654260 -79.360636                    Regent Park, Harbourfront  
1  43.662301 -79.389494  Queen's Park, Ontario Provincial Government  
2  43.657162 -79.378937                     Garden District, Ryerson  
3  43.651494 -79.375418                               St. James Town  
4  43.676357 -79.293031                                  The Beaches

In [120]:
toronto_data.drop('Neighbourhood', axis=1, inplace=True)

/opt/conda/envs/Python-3.7-main/lib/python3.7/site-packages/pandas/core/frame.py:3997: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  errors=errors,


In [121]:
toronto_data.head()

Postal Code           Borough   Latitude  Longitude  \
0         M5A  Downtown Toronto  43.654260 -79.360636   
1         M7A  Downtown Toronto  43.662301 -79.389494   
2         M5B  Downtown Toronto  43.657162 -79.378937   
3         M5C  Downtown Toronto  43.651494 -79.375418   
4         M4E      East Toronto  43.676357 -79.293031   

                                  Neighborhood  
0                    Regent Park, Harbourfront  
1  Queen's Park, Ontario Provincial Government  
2                     Garden District, Ryerson  
3                               St. James Town  
4                                  The Beaches

In [127]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = toronto_data

# merge manhattan_grouped with manhattan_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head() # check the last columns!

Postal Code           Borough   Latitude  Longitude  \
0         M5A  Downtown Toronto  43.654260 -79.360636   
1         M7A  Downtown Toronto  43.662301 -79.389494   
2         M5B  Downtown Toronto  43.657162 -79.378937   
3         M5C  Downtown Toronto  43.651494 -79.375418   
4         M4E      East Toronto  43.676357 -79.293031   

                                  Neighborhood  Cluster Labels  \
0                    Regent Park, Harbourfront               0   
1  Queen's Park, Ontario Provincial Government               0   
2                     Garden District, Ryerson               0   
3                               St. James Town               0   
4                                  The Beaches               0   

  1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0           Coffee Shop                   Pub                  Park   
1           Coffee Shop           Yoga Studio            Hobby Shop   
2           Coffee Shop        Clothing Store                  Café   
3           Coffee Shop                  Café          Cocktail Bar   
4     Health Food Store                   Pub                 Trail   

  4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
0                Bakery                  Café               Theater   
1            Restaurant        Sandwich Place   Distribution Center   
2   Japanese Restaurant       Bubble Tea Shop        Cosmetics Shop   
3            Restaurant             Gastropub   American Restaurant   
4               Dog Run          Dessert Shop    Dim Sum Restaurant   

       7th Most Common Venue   8th Most Common Venue 9th Most Common Venue  \
0             Breakfast Spot                 Brewery                   Spa   
1                      Diner      Italian Restaurant              Beer Bar   
2  Middle Eastern Restaurant  Furniture / Home Store        Lingerie Store   
3                   Beer Bar          Clothing Store                   Gym   
4                      Diner          Discount Store   Distribution Center   

  10th Most Common Venue  
0             Beer Store  
1                   Café  
2            Pizza Place  
3         Cosmetics Shop  
4          Women's Store

### Visualize the Clusters

In [129]:
toronto_merged.head(20)

Postal Code           Borough   Latitude  Longitude  \
0          M5A  Downtown Toronto  43.654260 -79.360636   
1          M7A  Downtown Toronto  43.662301 -79.389494   
2          M5B  Downtown Toronto  43.657162 -79.378937   
3          M5C  Downtown Toronto  43.651494 -79.375418   
4          M4E      East Toronto  43.676357 -79.293031   
5          M5E  Downtown Toronto  43.644771 -79.373306   
6          M5G  Downtown Toronto  43.657952 -79.387383   
7          M6G  Downtown Toronto  43.669542 -79.422564   
8          M5H  Downtown Toronto  43.650571 -79.384568   
9          M6H      West Toronto  43.669005 -79.442259   
10         M5J  Downtown Toronto  43.640816 -79.381752   
11         M6J      West Toronto  43.647927 -79.419750   
12         M4K      East Toronto  43.679557 -79.352188   
13         M5K  Downtown Toronto  43.647177 -79.381576   
14         M6K      West Toronto  43.636847 -79.428191   
15         M4L      East Toronto  43.668999 -79.315572   
16         M5L  Downtown Toronto  43.648198 -79.379817   
17         M4M      East Toronto  43.659526 -79.340923   
18         M4N   Central Toronto  43.728020 -79.388790   
19         M5N   Central Toronto  43.711695 -79.416936   

                                         Neighborhood  Cluster Labels  \
0                           Regent Park, Harbourfront               0   
1         Queen's Park, Ontario Provincial Government               0   
2                            Garden District, Ryerson               0   
3                                      St. James Town               0   
4                                         The Beaches               0   
5                                         Berczy Park               0   
6                                  Central Bay Street               0   
7                                            Christie               0   
8                            Richmond, Adelaide, King               0   
9                        Dufferin, Dovercourt Village               0   
10  Harbourfront East, Union Station, Toronto Islands               0   
11                           Little Portugal, Trinity               0   
12                       The Danforth West, Riverdale               0   
13           Toronto Dominion Centre, Design Exchange               0   
14       Brockton, Parkdale Village, Exhibition Place               0   
15                     India Bazaar, The Beaches West               0   
16                     Commerce Court, Victoria Hotel               0   
17                                    Studio District               0   
18                                      Lawrence Park               2   
19                                           Roselawn               4   

   1st Most Common Venue 2nd Most Common Venue  3rd Most Common Venue  \
0            Coffee Shop                   Pub                   Park   
1            Coffee Shop           Yoga Studio             Hobby Shop   
2            Coffee Shop        Clothing Store                   Café   
3            Coffee Shop                  Café           Cocktail Bar   
4      Health Food Store                   Pub                  Trail   
5            Coffee Shop        Farmers Market                 Bakery   
6            Coffee Shop                  Café         Sandwich Place   
7          Grocery Store                  Café                   Park   
8            Coffee Shop                  Café                  Hotel   
9               Pharmacy                Bakery          Grocery Store   
10           Coffee Shop              Aquarium                  Hotel   
11                   Bar           Coffee Shop  Vietnamese Restaurant   
12      Greek Restaurant           Coffee Shop     Italian Restaurant   
13           Coffee Shop                 Hotel             Restaurant   
14                  Café           Coffee Shop         Breakfast Spot   
15                  Park        Sandwich Place   Fast Food Restaurant   
16       

In [131]:
# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

In [132]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

### 5. Examine Clusters¶

##### Cluster 1

In [133]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels          1st Most Common Venue  \
0   Downtown Toronto               0                    Coffee Shop   
1   Downtown Toronto               0                    Coffee Shop   
2   Downtown Toronto               0                    Coffee Shop   
3   Downtown Toronto               0                    Coffee Shop   
4       East Toronto               0              Health Food Store   
5   Downtown Toronto               0                    Coffee Shop   
6   Downtown Toronto               0                    Coffee Shop   
7   Downtown Toronto               0                  Grocery Store   
8   Downtown Toronto               0                    Coffee Shop   
9       West Toronto               0                       Pharmacy   
10  Downtown Toronto               0                    Coffee Shop   
11      West Toronto               0                            Bar   
12      East Toronto               0               Greek Restaurant   
13  Downtown Toronto               0                    Coffee Shop   
14      West Toronto               0                           Café   
15      East Toronto               0                           Park   
16  Downtown Toronto               0                    Coffee Shop   
17      East Toronto               0                    Coffee Shop   
20   Central Toronto               0                           Park   
22      West Toronto               0             Mexican Restaurant   
23   Central Toronto               0                    Coffee Shop   
24   Central Toronto               0                 Sandwich Place   
25      West Toronto               0                      Gift Shop   
26   Central Toronto               0                    Pizza Place   
27  Downtown Toronto               0                           Café   
28      West Toronto               0                           Café   
30  Downtown Toronto               0  Vegetarian / Vegan Restaurant   
31   Central Toronto               0                    Coffee Shop   
32  Downtown Toronto               0                 Airport Lounge   
34  Downtown Toronto               0                    Coffee Shop   
35  Downtown Toronto               0                    Coffee Shop   
36  Downtown Toronto               0                    Coffee Shop   
37  Downtown Toronto               0                    Coffee Shop   
38      East Toronto               0                     Skate Park   

   2nd Most Common Venue  3rd Most Common Venue 4th Most Common Venue  \
0                    Pub                   Park                Bakery   
1            Yoga Studio             Hobby Shop            Restaurant   
2         Clothing Store                   Café   Japanese Restaurant   
3                   Café           Cocktail Bar            Restaurant   
4                    Pub                  Trail               Dog Run   
5         Farmers Market                 Bakery          Cocktail Bar   
6                   Café         Sandwich Place    Italian Restaurant   
7                   Café                   Park           Candy Store   
8                   Café                  Hotel            Restaurant   
9                 Bakery          Grocery Store               Brewery   
10              Aquarium                  Hotel                  Café   
11           Coffee Shop  Vietnamese Restaurant            Restaurant   
12           Coffee Shop     Italian Restaurant            Restaurant   
13                 Hotel             Restaurant                  Café   
14           Coffee Shop         Breakfast Spot             Nightclub   
15        Sandwich Place   Fast Food Restaurant     Fish & Chips Shop   
16            Restaurant                  Hotel                  Café   
17                Bakery              Gastropub               Brewery   
20                 Hotel         Breakfast Spot        Sandwich Place   
22                  Café                    Bar       Thai Restaura

##### Cluster 2

In [134]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
29  Central Toronto               1                 Trail   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
29            Playground         Women's Store          Dance Studio   

          5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
29  Eastern European Restaurant   Dumpling Restaurant            Donut Shop   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
29      Doner Restaurant               Dog Run    Distribution Center

##### Cluster 3

In [135]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
18  Central Toronto               2                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
18              Bus Line           Swim School    College Rec Center   

   5th Most Common Venue        6th Most Common Venue 7th Most Common Venue  \
18      Department Store  Eastern European Restaurant   Dumpling Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
18            Donut Shop      Doner Restaurant                Dog Run

##### Cluster 4

In [137]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
21   Central Toronto               3                  Park   
33  Downtown Toronto               3                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
21      Sushi Restaurant         Jewelry Store                 Trail   
33                 Trail            Playground          Dance Studio   

          5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
21  Eastern European Restaurant   Dumpling Restaurant            Donut Shop   
33  Eastern European Restaurant   Dumpling Restaurant            Donut Shop   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
21      Doner Restaurant               Dog Run    Distribution Center  
33      Doner Restaurant               Dog Run    Distribution Center

##### Cluster 5

In [138]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
19  Central Toronto               4           Music Venue   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
19                Garden         Women's Store         Deli / Bodega   

   5th Most Common Venue        6th Most Common Venue 7th Most Common Venue  \
19     Electronics Store  Eastern European Restaurant   Dumpling Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
19            Donut Shop      Doner Restaurant                Dog Run

# Interpretation Clusters

##### Clusters Toronto

* Cluster 1: Coffeeshop + Café + Restaurants
* Cluster 2: Trail
* Cluster 3: Park + Busline
* Cluster 4: Park + Restaurant
* Cluster 5: Music 